sql-plugin/src/main/scala/com/nvidia/spark/rapids/io/async/TrafficController.scala

```diff
@@ -85,38 +86,55 @@ class HostMemoryThrottle(val maxInFlightHostMemoryBytes: Long) extends Throttle
  *
  * This class is thread-safe as it is used by multiple tasks.
  */
-class TrafficController protected[rapids] (throttle: Throttle) {
+class TrafficController protected[rapids] (@GuardedBy("lock") throttle: Throttle) {
 
-  @GuardedBy("this")
+  @GuardedBy("lock")
   private var numTasks: Int = 0
 
+  private val lock = new ReentrantLock()
+  private val condition = lock.newCondition()
```

Comment:
Consider renaming to something more specific
```suggestion
  private val canBeScheduled = lock.newCondition()
```